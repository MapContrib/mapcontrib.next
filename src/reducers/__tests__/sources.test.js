import {
  initialState,
  default as reducer,
  SourceException
} from 'reducers/sources';

import {
  addSource,
  addFeaturesToSourceById,
  removeSourceById
} from 'actions/sources';

const defaultSourceAttributes = {
  id: 'id',
  name: 'name',
  origin: 'origin',
  features: []
};

const overpassSource = {
  ...defaultSourceAttributes,
  id: 'UUID_OVERPASS'
};

const osmoseSource = {
  ...defaultSourceAttributes,
  id: 'UUID_OSMOSE'
};

describe('Sources reducer', () => {
  it('Should return the initial state', () => {
    expect(reducer()).toBe(initialState);
  });

  it('Should be able to add a source', () => {
    const state = reducer(initialState, addSource(overpassSource));

    expect(state).toMatchSnapshot();

    const state2 = reducer(state, addSource(osmoseSource));

    expect(state2).toMatchSnapshot();
    expect(Object.keys(state2).length).toBe(2);
    expect(state2).not.toBe(state);
  });

  it('Should be able to remove a source', () => {
    const state = reducer(initialState, addSource(overpassSource));
    const state2 = reducer(state, addSource(osmoseSource));

    const state3 = reducer(state2, removeSourceById(overpassSource.id));

    expect(Object.keys(state3).length).toBe(1);
    expect(state3).toMatchSnapshot();
    expect(state3).not.toBe(state2);
  });

  it('Should be able to add a feature to a source', () => {
    const state = reducer(initialState, addSource(overpassSource));

    const state2 = reducer(
      state,
      addFeaturesToSourceById('UUID_OVERPASS', [{ id: 123, type: 'node' }])
    );

    expect(state2).toMatchSnapshot();
    expect(state2.UUID_OVERPASS.features).toHaveLength(1);
    expect(state2).not.toBe(state);
  });

  it('Should throw an exception when adding features to an unknown source', () => {
    const state = reducer(initialState, addSource(overpassSource));
    const addFeatures = () =>
      reducer(state, addFeaturesToSourceById('Unknown_ID', [{}]));

    expect(addFeatures).toThrow(SourceException);
  });
});
