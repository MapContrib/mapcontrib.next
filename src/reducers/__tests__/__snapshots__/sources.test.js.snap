// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sources reducer Should be able to add a feature to a source 1`] = `
Object {
  "UUID_OVERPASS": Array [
    Object {
      "id": 123,
      "type": "node",
    },
  ],
}
`;

exports[`Sources reducer Should be able to add a source 1`] = `
Object {
  "UUID_OVERPASS": Object {
    "features": Array [],
    "id": "UUID_OVERPASS",
    "name": "name",
    "origin": "origin",
  },
}
`;

exports[`Sources reducer Should be able to add a source 2`] = `
Object {
  "UUID_OSMOSE": Object {
    "features": Array [],
    "id": "UUID_OSMOSE",
    "name": "name",
    "origin": "origin",
  },
  "UUID_OVERPASS": Object {
    "features": Array [],
    "id": "UUID_OVERPASS",
    "name": "name",
    "origin": "origin",
  },
}
`;

exports[`Sources reducer Should be able to remove a source 1`] = `
Object {
  "UUID_OSMOSE": Object {
    "features": Array [],
    "id": "UUID_OSMOSE",
    "name": "name",
    "origin": "origin",
  },
}
`;
